.github/workflows/cypress-test.yml

```diff
@@ -0,0 +1,131 @@
+name: Cypress Tests
+
+on: [push, pull_request]
+
+env:
+  TEST_BROWSER_HEADLESS: 1
+  CI: 1
+  FTR_PATH: 'ftr'
+  START_CMD: 'node ../scripts/opensearch_dashboards --dev --no-base-path --no-watch --opensearch_security.multitenancy.enable_aggregation_view=true'
+  OPENSEARCH_SNAPSHOT_CMD: 'node ../scripts/opensearch snapshot'
+  SPEC: 'cypress/integration/plugins/security-dashboard-plugin/aggregation_view.js,'
+
+jobs:
+  tests:
```

Comment:
This new tasks copies much of the logic from [integration-test.yml](https://github.com/opensearch-project/security-dashboards-plugin/blob/main/.github/workflows/integration-test.yml) we should reuse the workflow to make sure bugs/features are only added in one place and then diverge for the different test systems

Reply:
I agree the tests should not run one after the other in series and there are different configurations/requirements for both kinds of result.  Getting the customization you are after and benefit from reuse should still be possible with GitHub Actions syntax

> A workflow run is made up of one or more jobs, which run in parallel by default. To run jobs sequentially, you can define dependencies on other jobs using the jobs.<job_id>.needs keyword.

_From GitHub Actions Workflow syntax [1]_

I agree that we need different kinds of configuration and setup, but we should avoid some of that really tricky common code,  I think you could restructure the workflow with three jobs, maybe it will take more than three.  

- setup-dashboard-security
- setup-and-run-jest-tests (Needs setup-dashboard-security)
- setup-and-run-cypress-tests (Needs setup-dashboard-security)

GitHub Actions has another section on reusing workflows [2] which might be what you'd want or its total overkill.

A quick final note; it can be a pain to test GitHub Actions because you have to commit/push/wait, you might be able to use ACT [3] to execute the workflows locally while developing/testing

[1] https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions
[2] https://docs.github.com/en/actions/using-workflows/reusing-workflows
[3] https://github.com/nektos/act

Reply:
@peternied There is no need to add these workflow sequentially as they do not have dependencies on each other at all. They should run in parallel to avoid add more run-time to every integration test run. No matter how you want to add this flow to the existing integration tests(re-use or not), it would add more time to integration tests run.
There is no tricky common code. Yes, as you said, it would be total overkill to combine these 2 workflows.
@RyanL1997 has clearly explained why this workflow is separate. 
> The cypress requires a different [set of configurations](https://github.com/opensearch-project/security-dashboards-plugin/pull/1146/files#diff-1c29b5bd518bbbdff17cb10f614dd430171102ccd8c78497bebe5da820657dbbR101-R111) (e.g feature flags for aggregation view).
This workflow will take 16+ mins as it requires wait for [API calls](https://github.com/opensearch-project/opensearch-dashboards-functional-test/blob/def17c61cd928794c1b8af368e029d07d8c15f9f/cypress/utils/plugins/security-dashboards-plugin/commands.js) in cypress test itself. It is necessary to run this in parallel with the integration tests.
The workflow setup with specific configurations also requires longer time. ([our case](https://github.com/opensearch-project/security-dashboards-plugin/pull/1146/files#diff-1c29b5bd518bbbdff17cb10f614dd430171102ccd8c78497bebe5da820657dbbR113) and [other example](https://github.com/opensearch-project/index-management-dashboards-plugin/blob/main/.github/workflows/cypress-workflow.yml#L74)).
Running these tests separately will make the debug easier.

Reply:
@peternied There is no need to run this workflow and integration tests workflow sequentially as they do not have dependencies on each other at all. They should run in parallel to avoid adding more run-time to every integration test run. No matter how to add this flow to the existing integration tests(re-use or not), that would add more time to integration tests run.
There is no tricky common code. Yes, as you said, it would be total overkill to combine these 2 workflows.
@RyanL1997 has explained why this workflow is separate.

The cypress requires a different [set of configurations](https://github.com/opensearch-project/security-dashboards-plugin/pull/1146/files#diff-1c29b5bd518bbbdff17cb10f614dd430171102ccd8c78497bebe5da820657dbbR101-R111) (e.g feature flags for aggregation view).
This workflow will take 16+ mins as it requires wait for [API calls](https://github.com/opensearch-project/opensearch-dashboards-functional-test/blob/def17c61cd928794c1b8af368e029d07d8c15f9f/cypress/utils/plugins/security-dashboards-plugin/commands.js) in cypress test itself. It is necessary to run this in parallel with the integration tests.
The workflow setup with specific configurations also requires longer time. ([our case](https://github.com/opensearch-project/security-dashboards-plugin/pull/1146/files#diff-1c29b5bd518bbbdff17cb10f614dd430171102ccd8c78497bebe5da820657dbbR113) and [other example](https://github.com/opensearch-project/index-management-dashboards-plugin/blob/main/.github/workflows/cypress-workflow.yml#L74)).
Running these tests separately will make the debug easier.

Reply:
> There is no tricky common code.

I disagree with this, there are ~70 lines of complex code in these workflows, [1] [2].  This can be separated into multiple jobs in the same file and referenced by a jest and cypress jobs that run in parallel.  Are there other areas of concern that I am missing?

#### Job reuse vs workflow reuse
While at the job level and workflow level reuse is possible in GitHub actions.  I think job reuse is appropriate for this repo, but workflow reuse is not needed.  A good candidate for workflow sharing would be for Dashboards or Dashboard-Plugins project to have workflow that prepped and set itself up so plugins like this one could install themselves, configure and executes testing.  The Security-Dashboards-plugin doesn't have downstream consumers like this.


[1] https://github.com/opensearch-project/security-dashboards-plugin/blob/main_aggregation_view/.github/workflows/cypress-test.yml#L18-L94
[2] https://github.com/opensearch-project/security-dashboards-plugin/blob/main_aggregation_view/.github/workflows/integration-test.yml#L14-L100

Reply:
@peternied Github does not have reusable jobs at all. It has reusable workflow instead. But in order to re-use those setup steps, the steps need to be extracted and to be put in a separate workflow. That would need to be done in a separate PR if needed. 

In addition, The setup steps are done separately not only in this Cypress workflow, but also in the [unit tests workflow](https://github.com/opensearch-project/security-dashboards-plugin/blob/main/.github/workflows/unit-test.yml#L11-L53).

Reply:
I think the lack of workflow reuse is an issue that is getting worse with this pull request.  I get how it seems like an obstacle for this feature rather than a benefit to the codebase that enables this testing this feature.  Could you file an issue and then I'll resolve this thread?

Reply:
@peternied Here we go: https://github.com/opensearch-project/security-dashboards-plugin/issues/1156